File: src/core/schema/tools/print.rb
# This is a *generic* printing function, that can print a indented dump
# of any object. It uses the model to 

# obj: the object to be printed
# indent: amount of indent

module Print
  
  def self.print(obj)
    PrintC.print(obj)
  end

  class PrintC
    def initialize(output = $stdout, depth = nil)
      @output = output
      @depth = depth
    end

    def self.print(obj, depth = nil)
      self.new($stdout, depth).print(obj)
    end
  
    def self.to_s(obj, depth = nil)
      output = ""
      self.new(output, depth).print(obj)
      output
    end
      
    def print(obj, indent=0, back_link=nil)
      if !obj.respond_to?(:schema_class) then
        @output << "#{obj}\n"
      elsif obj.nil?
        @output << "nil\n"
      else
        klass = obj.schema_class   # TODO: pass as an argument for partial evaluation
        @output << "#{klass.name} #{obj.identity}\n"
        indent += 2
        klass.fields.each do |field|
          if field != back_link
            if field.type.is_a?("Primitive")
              data = (field.type.name == "str") ? "\"#{obj[field.name]}\"" : obj[field.name]
              @output << "#{' '.repeat(indent)}#{field.name}: #{data}\n"
            else
              if !field.many
                sub = obj[field.name]
                @output << "#{' '.repeat(indent)}#{field.name}: "
                if @depth && indent > @depth*2
                  @output << "...\n"
                else
                  print1(field.traversal, sub, indent, field.inverse)
                end
              elsif !obj[field.name].empty?
                @output << "#{' '.repeat(indent)}#{field.name}"
                subindent = indent + 2
                if @depth && indent > @depth*2
                  @output << " ...\n"
                else
                  @output << "\n"
                  obj[field.name].each_with_index do |sub, i|
                    @output << "#{' '.repeat(subindent)}##{i} "
                    print1(field.traversal, sub, subindent, field.inverse)
                  end
                end
              end
            end
          end
        end
      end
    end
    
    def print1(traversal, obj, indent, back_link)
      if obj.nil?
        @output << "nil\n"
      elsif traversal  
        print(obj, indent, back_link)
      else
        @output << "#{obj._path}\n"
      end
    end
  end
end

File: src/js/core/schema/tools/print.js
'use strict'

//// Print ////

var cwd = process.cwd() + '/';

var Print;

var print = function(obj) {
  var self = this;
  return PrintC.print(obj);
};

class PrintC {
  static new(...args) { return new PrintC(...args) };

  static print(obj, depth = null) {
    var self = this;
    return self.new(Enso.System.stdout(), depth).print(obj);
  };

  static to_s(obj, depth = null) {
    var self = this, output;
    output = "";
    self.new(output, depth).print(obj);
    return output;
  };

  constructor(output = Enso.System.stdout(), depth = null) {
    var self = this;
    self.output$ = output;
    self.depth$ = depth;
  };

  print(obj, indent = 0, back_link = null) {
    var self = this, klass, data, sub, subindent;
    if (! obj.respond_to_P("schema_class")) {
      return self.output$.push(Enso.S(obj, "\n"));
    } else if (obj == null) {
      return self.output$.push("nil\n");
    } else {
      klass = obj.schema_class();
      self.output$.push(Enso.S(klass.name(), " ", obj.identity(), "\n"));
      indent = indent + 2;
      return klass.fields().each(function(field) {
        if (field != back_link) {
          if (Enso.System.test_type(field.type(), "Primitive")) {
            data = field.type().name() == "str"
              ? Enso.S("\"", obj.get$(field.name()), "\"")
              : obj.get$(field.name());
            return self.output$.push(Enso.S(" ".repeat(indent), field.name(), ": ", data, "\n"));
          } else if (! field.many()) {
            sub = obj.get$(field.name());
            self.output$.push(Enso.S(" ".repeat(indent), field.name(), ": "));
            if (self.depth$ && indent > self.depth$ * 2) {
              return self.output$.push("...\n");
            } else {
              return self.print1(field.traversal(), sub, indent, field.inverse());
            }
          } else if (! obj.get$(field.name()).empty_P()) {
            self.output$.push(Enso.S(" ".repeat(indent), field.name()));
            subindent = indent + 2;
            if (self.depth$ && indent > self.depth$ * 2) {
              return self.output$.push(" ...\n");
            } else {
              self.output$.push("\n");
              return obj.get$(field.name()).each_with_index(function(sub, i) {
                self.output$.push(Enso.S(" ".repeat(subindent), "#", i, " "));
                return self.print1(field.traversal(), sub, subindent, field.inverse());
              });
            }
          }
        }
      });
    }
  };

  print1(traversal, obj, indent, back_link) {
    var self = this;
    if (obj == null) {
      return self.output$.push("nil\n");
    } else if (traversal) {
      return self.print(obj, indent, back_link);
    } else {
      return self.output$.push(Enso.S(obj._path(), "\n"));
    }
  };
};

Print = {
  print: print,
  PrintC: PrintC,
};
module.exports = Print ;
